### PR TITLE
Correct tag highlight in Log Graph

### PR DESCRIPTION
**Pull request by gollth**
## Problem
Currently if a commit contains a tag, the tag will be rendered in the log graph with the `NeogitBranch` highlight group. Worse, though, it can happen, that the tag will be misinterpreted as the remote name:

| Before | After |
|:--|:--|
|graph-tags-before|graph-tags-after|

## Tech Details

The `commit.ref_name` might also contain a tag name, which the log graph is currently rendering in the `NeogitBranch` highlight group. The refname of HEAD with a tag placed on it might look like this:

> HEAD -> main, tag: 1.0.0, origin/main, origin/HEAD

A refname of a tag not associated to any branch might look like this:

> tag: 1.0.0

This makes the parsing of `local_name`, `remote_name` a bit more sophisticated, since we don't know the exact order of them. Naive solution might be to just split `ref_name` by comma and pattern match:

* contains `tag: ...` -> `tag_name`
* contains `/` -> `remote_name`
* otherwise -> `local_name`

No Lua expert, maybe there is a better way to solve this =)

### CODE REVIEW

**Comment by CKolkey** [comment]
The one sticky thing is that sometimes people (myself included) will use `/` in a branch name, so you wind up with a remote like `origin/fix/thing-112` and a local like `fix/thing-112`. I've used this to parse that out [elsewhere](https://github.com/NeogitOrg/neogit/blob/9316e40b915f92ea11102933630e799f58772f56/lua/neogit/popups/push/actions.lua#L55):

```lua
upstream = "origin/fix/thing-112"
remote, branch = upstream:match("^([^/]*)/(.*)$")
```

That pattern reads as:
- From the start `^`
- Capture everything that _isn't_ a "/" `([^/]*)`
- Ignore the first "/" `/`
- Capture everything else to the end `(.*)$`

Anyways, in practice, a better heuristic for the "local-vs-remote" problem is that a local branch has zero or more `/`'s, and a remote has one or more `/`'s. But that isn't great. So, you could use the above pattern to extract the "remote" name, and compare that to a list of remotes (`require("neogit.lib.git").remote.list()`). A little tedious, but should be the most correct.

If you wanna mess with that, go ahead. Otherwise this is still an improvement, so I'll take it as-is and try to find time myself

**Comment by gollth** [comment]
Thanks for the pointers @CKolkey, make sense to me. I try to implement a solution like you suggested over the next couple of days

**Comment by CKolkey** [comment]
No problem - I appreciate your attention to detail. Parsing CLI output has a seemingly never ending supply of edge cases.

**Comment by gollth** [comment]
Okay, turns out parsing all these combinations is quite finicky. I thought it make sense to unit test this behavior as well, so I moved it into `require("neogit.lib.git.log").interprete()` (if there is a better place for it, let me know).

This function uses now the output of `remotes.list()` like you suggested and it works quite robustly. For example when two remotes have the same branch on the same commit, they will now be rendered like this:

image

Don't know why some other tests are failing, on my local machine all is fine. Will investigate

**Review by CKolkey — CHANGES_REQUESTED** [review]
Only the most minor of nits to pick - otherwise looks great :)

File: lua/neogit/buffers/common.lua [hunk 1]
```diff
@@ -1,6 +1,8 @@
 local Ui = require("neogit.lib.ui")
 local Component = require("neogit.lib.ui.component")
 local util = require("neogit.lib.util")
+local log = require("neogit.lib.git.log")
+local remote = require("neogit.lib.git.remote")
```
[the remark this answers is not in the document]
Comment (gollth) [reply lua/neogit/buffers/common.lua]: Makes total sense. Done like you suggested

**Comment by gollth** [comment]
> Don't know why some other tests are failing, on my local machine all is fine. Will investigate

Okay so as it turns out, there was a problem with the ordering of where each branch is displayed on the graph. It would depend on the output of `git log ...`. Strangely this is not even consistent between git versions. The CI uses git 2.34, I assume, which produces different behavior than my local git 2.39 =S.

To be most robust, I implement a way to make it independent of the output of git. Now "untracked" branches (i.e. those who doesn't have a remote counterpart) are always listed first, then come the "tracked" ones. See 14dfa5d for an example. Impl isn't the prettiest, but it does the job.

**Comment by CKolkey** [comment]
I just merged your other branch, which made a little conflict 😅

**Comment by gollth** [comment]
Haha, no worries. Just rebased on latest master. Thanks for merging =)

**Comment by CKolkey** [comment]
Really great stuff here. Thanks for adding tests, too, there's a general... lack... of that. 

Easy merge, thanks for the contributions ;)
